services/server/src/handlers/availabilitySlot.ts

```diff
@@ -0,0 +1,205 @@
+import { bad, conflict, forbidden, notfound } from "@/lib/error";
+import { datetime, id, skippablePagination } from "@/validation/utils";
+import { isTutor, isTutorManager, isUser } from "@litespace/auth";
+import { IAvailabilitySlot } from "@litespace/types";
+import {
+  availabilitySlots,
+  interviews,
+  knex,
+  lessons,
+} from "@litespace/models";
+import dayjs from "@/lib/dayjs";
+import { NextFunction, Request, Response } from "express";
+import safeRequest from "express-async-handler";
+import { asSubSlots, isIntersecting } from "@litespace/sol";
+import zod from "zod";
+import { isEmpty } from "lodash";
+import { deleteSlots } from "@/lib/availabilitySlot";
+
+const findPayload = zod.object({
+  userId: id,
+  after: datetime,
+  before: datetime,
+  slotsOnly: zod.boolean().optional().default(false),
+  pagination: skippablePagination.optional(),
+});
+
+const setPayload = zod.object({
+  slots: zod.array(
+    zod.union([
+      zod.object({
+        action: zod.literal("create"),
+        start: datetime,
+        end: datetime,
+      }),
+      zod.object({
+        action: zod.literal("update"),
+        id: id,
+        start: datetime.optional(),
+        end: datetime.optional(),
+      }),
+      zod.object({
+        action: zod.literal("delete"),
+        id: id,
+      }),
+    ])
+  ),
+});
+
+async function find(req: Request, res: Response, next: NextFunction) {
+  const user = req.user;
+  if (!isUser(user)) return next(forbidden());
+
+  const { userId, after, before, slotsOnly, pagination } = findPayload.parse(
+    req.query
+  );
+
+  const diff = dayjs(before).diff(after, "days");
+  if (diff < 0) return next(bad());
+  const canUseFullFlag =
+    after && before && dayjs.utc(before).diff(after, "days") <= 30;
+
+  const paginatedSlots = await availabilitySlots.find({
+    users: [userId],
+    after,
+    before,
+    pagination: {
+      page: pagination?.page || 1,
+      size: pagination?.size || 10,
+      full: pagination?.full || !!canUseFullFlag,
+    },
+  });
+
+  // return only-slots only if the user is a tutor
+  if (slotsOnly && (isTutor(user.role) || isTutorManager(user.role))) {
+    const result = {
+      list: [{ slots: paginatedSlots.list, subslots: [] }],
+      total: paginatedSlots.total,
+    };
+    res.status(200).json(result);
+    return;
+  }
+
+  const slotIds = paginatedSlots.list.map((slot) => slot.id);
+
+  const paginatedLessons = await lessons.find({
+    users: [userId],
+    slots: slotIds,
+    after,
+    before,
+  });
+
+  const userInterviews = await interviews.find({
+    users: [userId],
+    slots: slotIds,
+  });
```

Comment:
We can move this logic into a seperate function. It will be more cleaner. Here is how the final code should look like. 

```ts
function getSubslot(payload: {slotIds: number[], userId: number}): Promise<IAvailabilitySlot.SubSlot[]>;

const subslots = isTutor(user) || isTutorManager(user) ? [] : getSubSlot(slotIds, user.id);
```

This way we will only have one `res.json()` and the logic will relatively simple to grasb what this handler is doing from a first look.